src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs

```diff
@@ -12,30 +12,49 @@ public class SocketReceiver
         private readonly Socket _socket;
         private readonly SocketAsyncEventArgs _eventArgs = new SocketAsyncEventArgs();
         private readonly SocketAwaitable _awaitable;
+        private Memory<byte> _buffer;
 
         public SocketReceiver(Socket socket, PipeScheduler scheduler)
         {
             _socket = socket;
             _awaitable = new SocketAwaitable(scheduler);
-            _eventArgs.UserToken = _awaitable;
-            _eventArgs.Completed += (_, e) => ((SocketAwaitable)e.UserToken).Complete(e.BytesTransferred, e.SocketError);
+            _eventArgs.SetBuffer(Array.Empty<byte>(), 0, 0);
+            _eventArgs.UserToken = this;
+            _eventArgs.Completed += (_, e) => ((SocketReceiver)e.UserToken).OnReadable();
         }
 
         public SocketAwaitable ReceiveAsync(Memory<byte> buffer)
         {
-#if NETCOREAPP2_1
-            _eventArgs.SetBuffer(buffer);
-#else
-            var segment = buffer.GetArray();
-
-            _eventArgs.SetBuffer(segment.Array, segment.Offset, segment.Count);
-#endif
+            _buffer = buffer;
             if (!_socket.ReceiveAsync(_eventArgs))
             {
-                _awaitable.Complete(_eventArgs.BytesTransferred, _eventArgs.SocketError);
+                OnReadable();
             }
-
             return _awaitable;
         }
+
+        private void OnReadable()
+        {
+            SocketError errorCode = _eventArgs.SocketError;
+            int bytesTransferred = 0;
+            if (errorCode == SocketError.Success)
+            {
+                try
+                {
+#if NETCOREAPP2_1
+                    bytesTransferred = _socket.Receive(_buffer.Span, SocketFlags.None, out errorCode);
+#else
+                    var segment = _buffer.GetArray();
+                    bytesTransferred = _socket.Receive(segment.Array, segment.Offset, segment.Count, SocketFlags.None, out errorCode);
+#endif
+                }
+                catch (ObjectDisposedException)
```

Comment:
Is this the only exception that could occur?

Reply:
Exceptions we get here are due to calling Dispose in SocketConnection. Errors for the receive operation are reported in errorCode.

Reply:
If we change the code in the way I suggest in my last comment, the ODE will be caught and converted into ConnectionAbortedException as it should without any new code.